packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2174,8 +2232,28 @@ export class KeyringController extends BaseController<
 
     const keyring = keyringBuilder();
 
-    // @ts-expect-error Enforce data type after updating clients
-    await keyring.deserialize(data);
+    // find the last index of the type
+    const lastIndexOfType: number = this.#keyrings.reduce((maxIndex, item) => {
+      if (item.type === type) {
+        return Math.max(
+          maxIndex,
+          (item as EthKeyring<Json> & { opts: { typeIndex: number } }).opts
+            ?.typeIndex ?? 0,
+        );
+      }
+      return maxIndex;
+    }, 0);
+
+    if (type === KeyringTypes.hd) {
+      await keyring.deserialize({
+        ...(data ?? {}),
+        typeIndex: lastIndexOfType + 1,
+        id: ulid(),
+      });
+    } else {
+      // @ts-expect-error Enforce data type after updating clients
+      await keyring.deserialize(data);
+    }
```

Comment:
If in methods like `addNewAccount` we allow a consumer to reference any kind of keyring by its ID, why we do this for the HD keyring only here?